car_control.py

```diff
@@ -1,137 +1,128 @@
-THROTTLE_SERVO_MAX = 0.25
-
-# Tweak these values for your robocar.
-STEERING_SERVO_MIN = -1.5
-STEERING_SERVO_MAX = 1.5
-
-###########
-# Setup
-###########
-
-MIXING_RATE = max(min(MIXING_RATE, 1.0), 0.0)
-
-THROTTLE_CUT_OFF_ANGLE = max(min(THROTTLE_CUT_OFF_ANGLE, 89.99), 0)
-THROTTLE_CUT_OFF_RATE = max(min(THROTTLE_CUT_OFF_RATE, 1.0), 0.01)
-
-THROTTLE_OFFSET = max(min(THROTTLE_OFFSET, 100), 0)
-STEERING_OFFSET = max(min(STEERING_OFFSET, 180), 0)
-
-# Handle if these were reversed...
-tmp = max(THROTTLE_SERVO_MIN, THROTTLE_SERVO_MAX)
-THROTTLE_SERVO_MIN = min(THROTTLE_SERVO_MIN, THROTTLE_SERVO_MAX)
-THROTTLE_SERVO_MAX = tmp
-
-# Handle if these were reversed...
-tmp = max(STEERING_SERVO_MIN, STEERING_SERVO_MAX)
-STEERING_SERVO_MIN = min(STEERING_SERVO_MIN, STEERING_SERVO_MAX)
-STEERING_SERVO_MAX = tmp
 
 class car_control:
-    old_cx_normal = None
-
-    throttle_old_result = None
-    throttle_i_output = 0
-    throttle_output = 0
-
-    steering_old_result = None
-    steering_i_output = 0
-    steering_output = STEERING_OFFSET
+    
+    def load(self, filename):
+        with open(filename, 'r') as filehandle:
+            jsonhandle = json.load(filehandle)
+        self.FRAME_WIDTH = float(jsonhandle["FRAME_WIDTH"])
+        self.MIXING_RATE = float(jsonhandle["MIXING_RATE"])                            # Percentage of a new line detection to mix into current steering.
+        self.THROTTLE_CUT_OFF_ANGLE = float(jsonhandle["THROTTLE_CUT_OFF_ANGLE"])      # Maximum angular distance from 90 before we cut speed [0.0-90.0).
+        self.THROTTLE_CUT_OFF_RATE = float(jsonhandle["THROTTLE_CUT_OFF_RATE"])        # How much to cut our speed boost (below) once the above is passed (0.0-1.0].
+        self.THROTTLE_GAIN = float(jsonhandle["THROTTLE_GAIN"])                        # e.g. how much to speed up on a straight away
+        self.THROTTLE_OFFSET = float(jsonhandle["THROTTLE_OFFSET"])                    # e.g. default speed (0 to 100)
+        self.THROTTLE_P_GAIN = float(jsonhandle["THROTTLE_P_GAIN"])
+        self.THROTTLE_I_GAIN = float(jsonhandle["THROTTLE_I_GAIN"])
+        self.THROTTLE_I_MIN = float(jsonhandle["THROTTLE_I_MIN"])
+        self.THROTTLE_I_MAX = float(jsonhandle["THROTTLE_I_MAX"])
+        self.THROTTLE_D_GAIN = float(jsonhandle["THROTTLE_D_GAIN"])
+        self.STEERING_OFFSET = float(jsonhandle["STEERING_OFFSET"])                    # Change this if you need to fix an imbalance in your car (0 to 180).
+        self.STEERING_P_GAIN = float(jsonhandle["STEERING_P_GAIN"])                    # Make this smaller as you increase your speed and vice versa.
+        self.STEERING_I_GAIN = float(jsonhandle["STEERING_I_GAIN"])
+        self.STEERING_I_MIN = float(jsonhandle["STEERING_I_MIN"])
+        self.STEERING_I_MAX = float(jsonhandle["STEERING_I_MAX"])
+        self.STEERING_D_GAIN = float(jsonhandle["STEERING_D_GAIN"])                    # Make this larger as you increase your speed and vice versa.
```

Comment:
Try not to go over 80 columns. Keep the original comments in addition to new comments for now.